src/app/code/community/Shopgate/Cloudapi/Helper/Request.php

```diff
@@ -60,14 +70,59 @@ protected function parameterDetected()
      */
     protected function setCookie()
     {
-        Mage::getSingleton('core/cookie')->set(self::COOKIE_NAME, self::COOKIE_VALUE, 0);
+        $data = new Varien_Object();
+        $data->setData(self::KEY_SGCLOUD_INAPP, self::COOKIE_VALUE);
+        $data->setData(self::KEY_SGCLOUD_IS_CHECKOUT, $this->getParam(self::KEY_SGCLOUD_IS_CHECKOUT));
+        $data->setData(self::KEY_SGCLOUD_CALLBACK_DATA, $this->getParam(self::KEY_SGCLOUD_CALLBACK_DATA));
+
+        Mage::getSingleton('core/cookie')->set(self::COOKIE_NAME, json_encode($data->getData()), 0);
     }
 
     /**
+     * @param string $key
+     * @param mixed  $value
      * @return bool
      */
-    protected function cookieIsSet()
+    public function cookieIsSet($key, $value = false)
+    {
+        $data = $this->getCookie();
+
+        if (!isset($data[$key])) {
+            return false;
+        }
+
+        if (!$value) {
+            return true;
+        } else {
+            return $data[$key] === $value;
+        }
+    }
+
+    /**
+     * @param string $key
+     * @return mixed
+     */
+    public function cookieGetValue($key)
+    {
+        $data = $this->getCookie();
+
+        return $data[$key];
```

Comment:
See comment above, NULL is also possible as a return of getCookie() and therefore $data[$key] will fail.